File: iosApp/iosApp/Features/Shop/ShopItemView.swift
//
//  ShopItemView.swift
//  BDE
//
//  Created by Nathan Fallet on 27/03/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared

struct ShopItemView: View {
    
    @EnvironmentObject var rootViewModel: RootViewModel
    @StateObject var viewModel: ShopItemViewModel
    
    var body: some View {
        ScrollView {
            VStack(alignment: .leading) {
                if let success = viewModel.success {
                    Text(success)
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .cardView(background: Color.green)
                }
                if let configuration = viewModel.item as? TicketConfiguration, Int64(truncating: configuration.userLeft ?? 1) < 1 {
                    Text("Cet élément n'est plus disponible.")
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .cardView(background: Color.accentColor)
                }
                ShopCard(
                    item: viewModel.item,
                    detailsEnabled: false,
                    cotisant: rootViewModel.user?.cotisant != nil
                )
                VStack(alignment: .leading, spacing: 8) {
                    HStack {
                        Text("Acheter")
                            .fontWeight(.bold)
                        Spacer()
                    }
                    ShopItemPrice(item: viewModel.item, cotisant: rootViewModel.user?.cotisant != nil)
                    if let configuration = viewModel.item as? TicketConfiguration, let userLeft = configuration.userLeft {
                        Text("\(userLeft) place(s) restante(s)")
                    }
                    Picker(selection: $viewModel.payNow, label: Text("")) {
                        Text("Lydia").tag(true)
                        Text("A un membre BDE").tag(false)
                    }
                    .pickerStyle(.segmented)
                    Button("Acheter") {
                        viewModel.launchBuy(token: rootViewModel.token)
                    }
                    .disabled(viewModel.loading)
                    .buttonStyle(FilledButtonStyle())
                }
                .cardView()
            }
            .padding()
        }
        .navigationTitle(Text(viewModel.item.title ?? "Boutique"))
        .onAppear(perform: viewModel.onAppear)
        .alert(isPresented: $viewModel.error) {
            Alert(
                title: Text("Une erreur est survenue !"),
                message: Text("Vérifiez que vous êtes bien connecté à internet, que cet élément est encore disponible et que vous ne l'avez pas déjà acheté.")
            )
        }
    }
    
}

File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/shop/ShopItemView.kt
package me.nathanfallet.bdeensisa.features.shop

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.AlertDialog
import androidx.compose.material.Button
import androidx.compose.material.Card
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.material.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import me.nathanfallet.bdeensisa.features.MainViewModel
import me.nathanfallet.bdeensisa.models.TicketConfiguration
import me.nathanfallet.bdeensisa.views.Picker

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun ShopItemView(
    modifier: Modifier = Modifier,
    viewModel: ShopItemViewModel,
    mainViewModel: MainViewModel
) {

    val user by mainViewModel.getUser().observeAsState()

    val payNow by viewModel.getPayNow().observeAsState()
    val loading by viewModel.getLoading().observeAsState()
    val error by viewModel.getError().observeAsState()
    val success by viewModel.getSuccess().observeAsState()

    LazyColumn(modifier) {
        stickyHeader {
            TopAppBar(
                title = { Text(text = viewModel.item.title ?: "Boutique") }
            )
        }
        item {
            Spacer(modifier = Modifier.height(8.dp))
        }
        if (success != null) {
            item {
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp)
                        .padding(vertical = 4.dp),
                    elevation = 4.dp,
                    backgroundColor = Color(0xFF0BDA51)
                ) {
                    Text(
                        text = success ?: "",
                        modifier = Modifier.padding(16.dp)
                    )
                }
            }
        }
        if (viewModel.item is TicketConfiguration && viewModel.item.userLeft != null && viewModel.item.userLeft!! < 1) {
            item {
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp)
                        .padding(vertical = 4.dp),
                    elevation = 4.dp,
                    backgroundColor = MaterialTheme.colors.primary
                ) {
                    Text(
                        text = "Cet élément n'est plus disponible.",
                        modifier = Modifier.padding(16.dp)
                    )
                }
            }
        }
        item {
            ShopCard(
                item = viewModel.item,
                detailsEnabled = false,
                cotisant = mainViewModel.getUser().value?.cotisant != null,
            )
        }
        item {
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp)
                    .padding(vertical = 4.dp),
                elevation = 4.dp
            ) {
                Column(
                    modifier = Modifier
                        .padding(16.dp)
                        .fillMaxWidth()
                ) {
                    Text(
                        text = "Acheter",
                        fontWeight = FontWeight.Bold
                    )
                    ShopItemPrice(viewModel.item, user?.cotisant != null)
                    if (viewModel.item is TicketConfiguration && viewModel.item.userLeft != null) {
                        Text(
                            text = "${viewModel.item.userLeft} place(s) restante(s)",
                            modifier = Modifier.padding(vertical = 8.dp)
                        )
                    }
                    Picker(
                        modifier = Modifier
                            .padding(vertical = 8.dp)
                            .fillMaxWidth(),
                        placeholder = "Méthode de paiement...",
                        items = mapOf(
                            true to "Lydia",
                            false to "A un membre BDE"
                        ),
                        selected = payNow ?: true,
                        onSelected = viewModel::setPayNow,
                    )
                    Button(
                        modifier = Modifier
                            .fillMaxWidth(),
                        enabled = loading != true,
                        onClick = {
                            viewModel.launchBuy(mainViewModel.getToken().value)
                        }
                    ) {
                        Text(text = "Acheter")
                    }
                    if (error == true) {
                        AlertDialog(
                            onDismissRequest = viewModel::dismissError,
                            title = { Text("Une erreur est survenue !") },
                            text = { Text("Vérifiez que vous êtes bien connecté à internet, que cet élément est encore disponible et que vous ne l'avez pas déjà acheté.") },
                            confirmButton = {
                                Button(onClick = viewModel::dismissError) {
                                    Text("OK")
                                }
                            }
                        )
                    }
                }
            }
        }
    }

}